grpc-context-utils/src/main/java/org/hypertrace/core/grpcutils/context/RequestContext.java

```diff
@@ -34,7 +35,7 @@ public class RequestContext {
   public static RequestContext forTenantId(String tenantId) {
     return new RequestContext()
         .put(RequestContextConstants.TENANT_ID_HEADER_KEY, tenantId)
-        .put(RequestContextConstants.REQUEST_ID_HEADER_KEY, UUID.randomUUID().toString());
+        .put(RequestContextConstants.REQUEST_ID_HEADER_KEY, UuidCreator.getRandomBasedFast().toString());
```

Comment:
Ref: https://github.com/f4b6a3/uuid-creator/blob/3b6471642e95ca91901ce5efe817537aa0d1068b/src/main/java/com/github/f4b6a3/uuid/UuidCreator.java#L174

Reply:
My concerns here aren't the new impl but
1. Adding a new dependency to a base lib like this impacts the dependency tree of every single service
2. This is just covering up an upstream problem - that we're generating request contexts inappropriately - many times for a single "request".

Reply:
>  Adding a new dependency to a base lib like this impacts the dependency tree of every single service

I can extract out the function from this library based on ThreadLocalRandom.

>This is just covering up an upstream problem - that we're generating request contexts inappropriately - many times for a single "request".

I didn't fully get this. Can you elaborate?

Reply:
> I didn't fully get this. Can you elaborate?

The reason this came up - at least the reason I'm aware of - as an issue is because ingester is generating many request contexts for a single trace ingestion. Each time it needs a context, it generates a new one. Because this now occurs across many enrichers and many threads in parallel, we're running into these performance problems. But not only is that unnecessary, it's actually wrong - we want the same ID to be shared for ingesting the same trace in the first place.

A couple different approaches there would be to either
1. update the enricher interface to accept the context, and generate it up front before we delegate the enrichment off to an async thread.
2. Create a request context from the trace details - use the trace id as the "request id". This would probably be defined in the ingestion pipeline repo.

Reply:
The request-id attribute is attached to the grpc-context while making a downstream call. This also seems to be getting added only for RXJava calls, if someone is not using RXJava, this request-id is not attached to the call.
How is this request-id used by the downstream services? Given that its attached only for RXJava based calls and not for all GRPC calls, do we even need to attach this request-id?

Reply:
>The reason this came up - at least the reason I'm aware of - as an issue is because ingester is generating many request contexts for a single trace ingestion. Each time it needs a context, it generates a new one. Because this now occurs across many enrichers and many threads in parallel, we're running into these performance problems. But not only is that unnecessary, it's actually wrong - we want the same ID to be shared for ingesting the same trace in the first place.

> A couple different approaches there would be to either

> update the enricher interface to accept the context, and generate it up front before we delegate the enrichment off to an async thread.
Create a request context from the trace details - use the trace id as the "request id". This would probably be defined in the ingestion pipeline repo.

I agree that for tracing the processing of StructureTrace, it's important to establish a context based on unique keys or attributes that can be used to generate a trace ID. And if `request_id` is primarily used internally to link outgoing requests to the same trace, we can consider option (2) having new request context based trace.

However, for any other consumers of this interface `forTenantId`, do we really need a `request_id` generated from a secure random source? Since this is primarily for tracing purposes and not for resource identification, I think it's more practical to use a fast UUID-based `request_id` generated with `ThreadLocalRandom`.

Reply:
> I can extract out the function from this library based on ThreadLocalRandom.

If we don't plan to use the above library here. I am thinking of having our own function like below:
```
public static UUID generateFastRandomUUID() {
    long mostSigBits = ThreadLocalRandom.current().nextLong();
    long leastSigBits = ThreadLocalRandom.current().nextLong();
    
    // Set the version (4) For random UUID 
    mostSigBits &= 0xFFFFFFFFFFFF0FFFL;
    mostSigBits |= 0x0000000000004000L;
    // Set variant to RFC 4122
    leastSigBits &= 0x3FFFFFFFFFFFFFFFL;  
    leastSigBits |= 0x8000000000000000L; 

    return new UUID(mostSigBits, leastSigBits);
}
```

Reply:
Few stacks showing up the contention at jdk level

```
"kafka-streams-async-worker-3" #97 daemon prio=5 os_prio=0 cpu=158592.16ms elapsed=573.69s tid=0x00007fdd9b4fb800 nid=0x570 waiting for monitor entry  [0x00007fdd6f422000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at sun.security.provider.NativePRNG$RandomIO.implNextBytes(java.base@11.0.19/Unknown Source)
	- waiting to lock <0x0000000115d07748> (a java.lang.Object)
	at sun.security.provider.NativePRNG.engineNextBytes(java.base@11.0.19/Unknown Source)
	at java.security.SecureRandom.nextBytes(java.base@11.0.19/Unknown Source)
	at java.util.UUID.randomUUID(java.base@11.0.19/Unknown Source)
	at org.hypertrace.core.grpcutils.context.RequestContext.forTenantId(RequestContext.java:37)
	at ai.traceable.platform.enrichment.clients.ApiAttributeOverrideCacheWithInvalidation.getApiIsExternalOverride(ApiAttributeOverrideCacheWithInvalidation.java:54)
	at ai.traceable.platform.apientity.ApiEntityDataAccessManager.evaluateIsExternalApi(ApiEntityDataAccessManager.java:499)
	at ai.traceable.platform.apientity.ApiEntityDataAccessManager.getUpdatedAttributes(ApiEntityDataAccessManager.java:253)
	at ai.traceable.platform.apientity.AbstractApiDataAccessManager.updateApiEntity(AbstractApiDataAccessManager.java:196)
	at ai.traceable.platform.apientity.http.discovery.HttpApiDiscoveryManager.getEntityForPattern(HttpApiDiscoveryManager.java:184)
	at ai.traceable.platform.apientity.http.discovery.HttpApiDiscoveryManager.getApiEntities(HttpApiDiscoveryManager.java:126)
	at ai.traceable.platform.enrichment.enrichers.ApiAttributeEnricher.enrichEvent(ApiAttributeEnricher.java:231)
	at org.hypertrace.traceenricher.enrichment.EnrichmentProcessor.applyEnricher(EnrichmentProcessor.java:123)
	at org.hypertrace.traceenricher.enrichment.EnrichmentProcessor.process(EnrichmentProcessor.java:81)
	at ai.traceable.platform.enrichment.enrichers.TraceableConvertingEnrichmentProcessor.enrichAndConvert(TraceableConvertingEnrichmentProcessor.java:38)
	at ai.traceable.platform.trace.enricher.StructuredTraceEnrichProcessor.asyncProcess(StructuredTraceEnrichProcessor.java:75)
	at ai.traceable.platform.trace.enricher.StructuredTraceEnrichProcessor.asyncProcess(StructuredTraceEnrichProcessor.java:26)
	at org.hypertrace.core.kafkastreams.framework.async.AsyncProcessor.lambda$process$0(AsyncProcessor.java:64)
	at org.hypertrace.core.kafkastreams.framework.async.AsyncProcessor$$Lambda$1077/0x00000008007df040.get(Unknown Source)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(java.base@11.0.19/Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.19/Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.19/Unknown Source)
	at java.lang.Thread.run(java.base@11.0.19/Unknown Source)
```


```
"kafka-streams-async-worker-15" #109 daemon prio=5 os_prio=0 cpu=162114.71ms elapsed=585.83s tid=0x00007fdd9b560000 nid=0x57c waiting for monitor entry  [0x00007fdd6dc16000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at sun.security.provider.NativePRNG$RandomIO.implNextBytes(java.base@11.0.19/Unknown Source)
	- waiting to lock <0x0000000115d07748> (a java.lang.Object)
	at sun.security.provider.NativePRNG.engineNextBytes(java.base@11.0.19/Unknown Source)
	at java.security.SecureRandom.nextBytes(java.base@11.0.19/Unknown Source)
	at java.util.UUID.randomUUID(java.base@11.0.19/Unknown Source)
	at org.hypertrace.core.grpcutils.context.RequestContext.forTenantId(RequestContext.java:37)
	at org.hypertrace.core.grpcutils.client.rx.GrpcRxExecutionContext.forTenantContext(GrpcRxExecutionContext.java:47)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.spanTenantContext(DefaultTraceEntityAccessor.java:197)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.resolveAllAttributes(DefaultTraceEntityAccessor.java:168)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.buildEntity(DefaultTraceEntityAccessor.java:125)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.writeEntityIfExists(DefaultTraceEntityAccessor.java:65)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.lambda$writeAssociatedEntitiesForSpanEventually$1(DefaultTraceEntityAccessor.java:61)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor$$Lambda$1518/0x0000000800b3f440.accept(Unknown Source)
	at java.util.ArrayList.forEach(java.base@11.0.19/Unknown Source)
	at org.hypertrace.trace.accessor.entities.DefaultTraceEntityAccessor.writeAssociatedEntitiesForSpanEventually(DefaultTraceEntityAccessor.java:61)
	at org.hypertrace.traceenricher.enrichment.enrichers.EntitySpanEnricher.enrichEvent(EntitySpanEnricher.java:19)
	at ai.traceable.platform.enrichment.enrichers.FeatureFlaggedEntitySpanEnricher.lambda$enrichEvent$2(FeatureFlaggedEntitySpanEnricher.java:35)
	at ai.traceable.platform.enrichment.enrichers.FeatureFlaggedEntitySpanEnricher$$Lambda$1510/0x0000000800b3cc40.accept(Unknown Source)
	at io.reactivex.rxjava3.internal.observers.BlockingMultiObserver.blockingConsume(BlockingMultiObserver.java:175)
	at io.reactivex.rxjava3.core.Maybe.blockingSubscribe(Maybe.java:2990)
	at io.reactivex.rxjava3.core.Maybe.blockingSubscribe(Maybe.java:2935)
	at ai.traceable.platform.enrichment.enrichers.FeatureFlaggedEntitySpanEnricher.enrichEvent(FeatureFlaggedEntitySpanEnricher.java:35)
	at org.hypertrace.traceenricher.enrichment.EnrichmentProcessor.applyEnricher(EnrichmentProcessor.java:123)
	at org.hypertrace.traceenricher.enrichment.EnrichmentProcessor.process(EnrichmentProcessor.java:81)
	at ai.traceable.platform.enrichment.enrichers.TraceableConvertingEnrichmentProcessor.enrichAndConvert(TraceableConvertingEnrichmentProcessor.java:38)
	at ai.traceable.platform.trace.enricher.StructuredTraceEnrichProcessor.asyncProcess(StructuredTraceEnrichProcessor.java:75)
	at ai.traceable.platform.trace.enricher.StructuredTraceEnrichProcessor.asyncProcess(StructuredTraceEnrichProcessor.java:26)
```